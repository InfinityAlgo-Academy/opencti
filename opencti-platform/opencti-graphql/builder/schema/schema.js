const esbuild = require('esbuild');
const {default: graphqlLoaderPlugin} = require('@luckycatfactory/esbuild-graphql-loader');

esbuild.build({
    logLevel: 'info',
    plugins: [graphqlLoaderPlugin()],
    entryPoints: ['script/script-generate-schema.js'],
    entryNames: "[name]",
    bundle: true,
    loader: {'.js': 'jsx'},
    platform: 'node',
    target: ['node14'],
    minify: true,
    keepNames: false,
    sourcemap: false,
    sourceRoot: 'src',
    sourcesContent: false,
    outdir: 'build',
    incremental: false,
});
