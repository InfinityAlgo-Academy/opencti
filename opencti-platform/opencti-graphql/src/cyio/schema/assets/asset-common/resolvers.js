import { assetSingularizeSchema as singularizeSchema, objectTypeMapping } from '../asset-mappings.js';
import {compareValues, updateQuery} from '../../utils.js';
import {
  getSelectSparqlQuery,
  getReducer,
  deleteMultipleAssetsQuery,
  removeMultipleAssetsFromInventoryQuery,
  deleteAssetQuery,
  removeAssetFromInventoryQuery,
  insertLocationQuery,
  selectLocationQuery,
  selectAllLocations,
  deleteLocationQuery,
  locationPredicateMap
} from './sparql-query.js';

const assetCommonResolvers = {
  Query: {
    assetList: async ( _, args, context, info  ) => { 
      var sparqlQuery = getSelectSparqlQuery('ASSET', );
      var reducer = getReducer('ASSET');
      const response = await context.dataSources.Stardog.queryAll( 
        context.dbName, 
        sparqlQuery,
        singularizeSchema,
        // args.first,       // limit
        // args.offset,      // offset
        args.filter       // filter
      )
      if (Array.isArray(response) && response.length > 0) {
        // build array of edges
        const edges = [];
        let limit = (args.first === undefined ? response.length : args.first) ;
        let offset = (args.offset === undefined ? 0 : args.offset) ;
        let assetList ;
        if (args.orderedBy !== undefined ) {
          assetList = response.sort(compareValues(args.orderedBy, args.orderMode ));
        } else {
          assetList = response;
        }

        // for each asset in the result set
        for (let asset of assetList) {
          // skip down past the offset
          if ( offset ) {
            offset--
            continue
          }

          // if haven't reached limit to be returned
          if ( limit ) {
            let edge = {
              cursor: asset.iri,
              node: reducer( asset ),
            }
            edges.push( edge )
            limit-- ;
          }
        }
        return {
          pageInfo: {
            startCursor: assetList[0].iri,
            endCursor: assetList[assetList.length -1 ].iri,
            hasNextPage: (args.first > assetList.length ? true : false),
            hasPreviousPage: (args.offset > 0 ? true : false),
            globalCount: assetList.length,
          },
          edges: edges,
        }
      } else {
        return [];
      }
    },
    asset: async ( _, args, context, info ) => {
      var sparqlQuery = getSelectSparqlQuery('ASSET', args.id);
      var reducer = getReducer('ASSET');
      const response = await context.dataSources.Stardog.queryById( context.dbName, sparqlQuery, singularizeSchema )
      if (response === undefined ) return null;
      const first = response[0];
      if (first === undefined) return null;
      return( reducer( first ) );
    },
    itAssetList: async ( _, args, context, info  ) => { 
      var sparqlQuery = getSelectSparqlQuery('IT-ASSET', );
      var reducer = getReducer('IT-ASSET');
      const response = await context.dataSources.Stardog.queryAll( 
        context.dbName, 
        sparqlQuery,
        singularizeSchema,
        // args.first,       // limit
        // args.offset,      // offset
        args.filter       // filter
      )
      if (Array.isArray(response) && response.length > 0) {
        // build array of edges
        const edges = [];
        let limit = (args.first === undefined ? response.length : args.first) ;
        let offset = (args.offset === undefined ? 0 : args.offset) ;
        let assetList ;
        if (args.orderedBy !== undefined ) {
          assetList = response.sort(compareValues(args.orderedBy, args.orderMode ));
        } else {
          assetList = response;
        }

        // for each asset in the result set
        for (let asset of assetList) {
          // skip down past the offset
          if ( offset ) {
            offset--
            continue
          }

          // if haven't reached limit to be returned
          if ( limit ) {
            let edge = {
              cursor: asset.iri,
              node: reducer( asset ),
            }
            edges.push( edge )
            limit-- ;
          }
        }
        return {
          pageInfo: {
            startCursor: assetList[0].iri,
            endCursor: assetList[assetList.length -1 ].iri,
            hasNextPage: (args.first > assetList.length),
            hasPreviousPage: (args.offset > 0),
            globalCount: assetList.length,
          },
          edges: edges,
        }
      } else {
        return [];
      }
    },
    itAsset: async ( _, args, context, info ) => {
      var sparqlQuery = getSelectSparqlQuery('IT-ASSET', args.id);
      var reducer = getReducer('IT-ASSET');
      const response = await context.dataSources.Stardog.queryById( context.dbName, sparqlQuery, singularizeSchema )
      if (response === undefined ) return null;
      const first = response[0];
      if (first === undefined) return null;
      return( reducer( first ) );
    },
    assetLocationList: async (_, args, context) => {
      const { dbName } = context;
      const query = selectAllLocations(context.selectMap.getNode("node"));
      const response = await context.dataSources.Stardog.queryAll(dbName, query, singularizeSchema, args.filter);
      if (Array.isArray(response) && response.length > 0) {
        const edges = [];
        const reducer = getReducer("ASSET-LOCATION");
        let limit = (args.first === undefined ? response.length : args.first) ;
        let offset = (args.offset === undefined ? 0 : args.offset) ;
        let locationList ;
        if (args.orderedBy !== undefined ) {
          locationList = response.sort(compareValues(args.orderedBy, args.orderMode ));
        } else {
          locationList = response;
        }

        // for each asset in the result set
        for (let location of locationList) {
          // skip down past the offset
          if (offset) {
            offset--
            continue
          }

          // if haven't reached limit to be returned
          if (limit) {
            let edge = {
              cursor: location.iri,
              node: reducer(location),
            }
            edges.push(edge)
            limit--;
          }
        }
        return {
          pageInfo: {
            startCursor: locationList[0].iri,
            endCursor: locationList[locationList.length - 1].iri,
            hasNextPage: (args.first > locationList.length),
            hasPreviousPage: (args.offset > 0),
            globalCount: locationList.length,
          },
          edges: edges,
        }
      } else {
        return [];
      }
    },
    assetLocation: async (_, {id}, context) => {
      const { dbName } = context;
      const query = selectLocationQuery(id, context.selectMap.getNode("assetLocation"));
      const response = await context.dataSources.Stardog.queryById(dbName, query, singularizeSchema);
      if(response === undefined || response.length === 0) return null;
      const reducer = getReducer("ASSET-LOCATION");
      return reducer(response[0]);
    }
  },
  Mutation: {
    deleteAsset: async (_, {id}, context) => {
      const dbName = context.dbName;
      const dq = deleteAssetQuery(id);
      await context.dataSources.Stardog.delete(dbName, dq);
      const ra = removeAssetFromInventoryQuery(id);
      await context.dataSources.Stardog.delete(dbName, ra);
    },
    deleteAssets: async (_, { ids }, context) => {
      const dbName = context.dbName;
      const dq = deleteMultipleAssetsQuery(ids);
      await context.dataSources.Stardog.delete(dbName, dq);
      const ra = removeMultipleAssetsFromInventoryQuery(ids);
      await context.dataSources.Stardog.delete(dbName, ra);
    },
    createAssetLocation: async (_, {input}, context) => {
      const { dbName } = context;
      const {id, query} = insertLocationQuery(input);
      await context.dataSources.Stardog.create(dbName, query);
      const select = selectLocationQuery(id, context.selectMap.getNode("createAssetLocation"));
      const result = await context.dataSources.Stardog.queryById(dbName, select, singularizeSchema);
      const reducer = getReducer("ASSET-LOCATION");
      return reducer(result[0]);
    },
    deleteAssetLocation: async (_, {id}, context) => {
      const { dbName } = context;
      const query = deleteLocationQuery(id);
      await context.dataSources.Stardog.delete(dbName, query);
      return id;
    },
    editAssetLocation: async (_, {id, input}, context) => {
      const { dbName } = context;
      const query = updateQuery(
          `http://darklight.ai/ns/common#CivicLocation-${id}`,
          "http://darklight.ai/ns/common#CivicLocation",
          input,
          locationPredicateMap
      )
      await context.dataSources.Stardog.edit(dbName, query);
      const select = selectLocationQuery(id, context.selectMap.getNode("editAssetLocation"));
      const result = await context.dataSources.Stardog.queryById(dbName, select, singularizeSchema);
      const reducer = getReducer("ASSET-LOCATION");
      return reducer(result[0]);
    }
  },
  // Map enum GraphQL values to data model required values
  AssetType: {
    account: 'account',
    appliance: 'appliance',
    application_software: 'application-software',
    circuit: 'circuit',
    computer_account: 'computer-account',
    compute_device: 'compute-device',
    data: 'data',
    database: 'database',
    directory_server: 'directory-server',
    dns_server: 'dns-server',
    email_server: 'email-server',
    embedded: 'embedded',
    firewall: 'firewall',
    guidance: 'guidance',
    hypervisor: 'hypervisor',
    load_balancer: 'load-balancer',
    network_device: 'network-device',
    network: 'network',
    operating_system: 'operating-system',
    pbx: 'pbx',
    physical_device: 'physical-device',
    plan: 'plan',
    policy: 'policy',
    printer: 'printer',
    procedure: 'procedure',
    router: 'router',
    server: 'server',
    service_account: 'service-account',
    service: 'service',
    software: 'software',
    standard: 'standard',
    storage_array: 'storage-array',
    switch: 'switch',
    system: 'system',
    user_account: 'user-account',
    validation: 'validation',
    voip_device: 'voip-device',
    voip_handset: 'voip-handset',
    voip_router: 'voip-router',
    web_server: 'web-server',
    web_site: 'web-site',
    workstation: 'workstation',
  },
  Asset: {
    __resolveType: ( item ) => {
      return objectTypeMapping[item.entity_type];
    },
    locations: ( parent, ) => {
    },
    external_references: ( parent, ) => {
    },
    notes: ( parent, ) => {
    },
  },
  AssetLocation: {
    __resolveType: ( item ) => {
      return objectTypeMapping[item.entity_type]
    }
  },
  HardwareAsset: {
    __resolveType: ( item ) => {
      return objectTypeMapping[item.entity_type];
    }
  },
  ItAsset: {
    __resolveType: ( item ) => {
      return objectTypeMapping[item.entity_type];
    }
  },
  AssetKind: {
    __resolveType: ( item ) => {
      return objectTypeMapping[item.entity_type];
    }
  },
  HardwareKind: {
    __resolveType: ( item ) => {
      return objectTypeMapping[item.entity_type];
    }
  },
  ItAssetKind: {
    __resolveType: ( item ) => {
      return objectTypeMapping[item.entity_type];
    }
  },
  IpAddress: {
    __resolveType: ( item ) => {
      return objectTypeMapping[item.entity_type];
    },
  },
  PortRange: {
    __resolveType: ( item ) => {
      return objectTypeMapping[item.entity_type];
    }
  },
};

export default assetCommonResolvers;
