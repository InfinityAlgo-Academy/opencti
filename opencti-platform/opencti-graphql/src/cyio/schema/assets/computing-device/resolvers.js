import { assetSingularizeSchema as singularizeSchema, objectTypeMapping } from '../asset-mappings.js';
import {getSelectSparqlQuery, getReducer, insertQuery, predicateMap} from './sparql-query.js';
import { getSelectSparqlQuery as getSoftwareQuery, 
         getReducer as getSoftwareReducer } from '../software/sparql-query.js';
import {compareValues, updateQuery} from '../../utils.js';
import {addToInventoryQuery, deleteQuery, removeFromInventoryQuery} from "../assetUtil";
import {
  deleteIpQuery,
  deletePortQuery,
  insertIPQuery,
  insertIPRelationship,
  insertPortRelationships,
  insertPortsQuery
} from "../assetQueries";
import {UserInputError} from "apollo-server-express";

const computingDeviceResolvers = {
  Query: {
    computingDeviceAssetList: async ( _, args, context, info  ) => { 
      var sparqlQuery = getSelectSparqlQuery('COMPUTING-DEVICE', );
      var reducer = getReducer('COMPUTING-DEVICE');
      const response = await context.dataSources.Stardog.queryAll( 
        context.dbName, 
        sparqlQuery,
        singularizeSchema,
        // args.first,       // limit
        // args.offset,      // offset
        args.filter       // filter
      )
      if (Array.isArray(response) && response.length > 0) {
        // build array of edges
        const edges = [];
        let limit = (args.first === undefined ? response.length : args.first) ;
        let offset = (args.offset === undefined ? 0 : args.offset) ;
        let assetList ;
        if (args.orderedBy !== undefined ) {
          assetList = response.sort(compareValues(args.orderedBy, args.orderMode ));
        } else {
          assetList = response;
        }

        // for each asset in the result set
        for (let asset of assetList) {
          // skip down past the offset
          if ( offset ) {
            offset--
            continue
          }

          // if haven't reached limit to be returned
          if ( limit ) {
            let edge = {
              cursor: asset.iri,
              node: reducer( asset ),
            }
            edges.push( edge )
            limit-- ;
          }
        }
        return {
          pageInfo: {
            startCursor: assetList[0].iri,
            endCursor: assetList[assetList.length -1 ].iri,
            hasNextPage: (args.first > assetList.length ? true : false),
            hasPreviousPage: (args.offset > 0 ? true : false),
            globalCount: assetList.length,
          },
          edges: edges,
        }
      } else {
        return;
      }
    },
    computingDeviceAsset: async ( _, args, context, info ) => {
      var sparqlQuery = getSelectSparqlQuery('COMPUTING-DEVICE', args.id);
      var reducer = getReducer('COMPUTING-DEVICE');
      const response = await context.dataSources.Stardog.queryById( context.dbName, sparqlQuery, singularizeSchema )
      if (response === undefined ) return null;
      const first = response[0];
      if (first === undefined) return null;
      return( reducer( first ) );
    },
  },
  Mutation: {
    createComputingDeviceAsset: async ( _, {input}, context ) => {
      const dbName = context.dbName;
      let ports, ipv4, ipv6;
      if(input.ports !== undefined) {
        ports = input.ports
        delete input.ports;
      }
      if(input.ipv4_address !== undefined) {
        ipv4 = input.ipv4_address;
        delete input.ipv4_address;
      }
      if(input.ipv6_address !== undefined) {
        ipv6 = input.ipv6_address;
        delete input.ipv6_address;
      }
      const {iri, id, query} = insertQuery(input);
      await context.dataSources.Stardog.create(dbName, query);
      const connectQuery = addToInventoryQuery(iri);
      await context.dataSources.Stardog.create(dbName, connectQuery);
      if(ports !== null){
        const {iris: portIris, query: portsQuery} = insertPortsQuery(ports);
        await context.dataSources.Stardog.create(dbName, portsQuery);
        const relationshipQuery = insertPortRelationships(iri, portIris);
        await context.dataSources.Stardog.create(dbName, relationshipQuery);
      }
      if(ipv4 !== null) {
        const {ipIris, query} = insertIPQuery(ipv4, 4);
        await context.dataSources.Stardog.create(dbName, query);
        const relationshipQuery = insertIPRelationship(iri, ipIris);
        await context.dataSources.Stardog.create(dbName, relationshipQuery);
      }
      if(ipv6 !== null) {
        const {ipIris, query} = insertIPQuery(ipv6, 6);
        await context.dataSources.Stardog.create(dbName, query);
        const relationshipQuery = insertIPRelationship(iri, ipIris);
        await context.dataSources.Stardog.create(dbName, relationshipQuery);
      }
      return {id};
    },
    deleteComputingDeviceAsset: async ( _, {id}, context, input ) => {
      const dbName = context.dbName;
      const sparqlQuery = getSelectSparqlQuery('COMPUTING-DEVICE', id);
      const response = await context.dataSources.Stardog.queryById( context.dbName, sparqlQuery, singularizeSchema )
      const reducer = getReducer('COMPUTING-DEVICE');
      if(response.length === 0) throw new UserInputError(`Entity does not exist with ID ${id}`);
      const asset = (reducer(response[0]));
      for(const portIri of asset.ports_iri) {
        const portQuery = deletePortQuery(portIri);
        await context.dataSources.Stardog.delete(dbName, portQuery);
      }
      for(const ipId of asset.ip_addr_iri) {
        const ipQuery = deleteIpQuery(ipId);
        await context.dataSources.Stardog.delete(dbName, ipQuery);
      }
      const relationshipQuery = removeFromInventoryQuery(id);
      await context.dataSources.Stardog.delete(dbName, relationshipQuery)
      const query = deleteQuery(id)
      await context.dataSources.Stardog.delete(dbName, query)
      return {id};
    },
    editComputingDeviceAsset: async ( _, {id, input}, context ) => {
      const dbName = context.dbName;
      const query = updateQuery(
          `http://scap.nist.gov/ns/asset-identification#ComputingDevice-${id}`,
          "http://scap.nist.gov/ns/asset-identification#ComputingDevice",
          input,
          predicateMap
      )
      await context.dataSources.Stardog.edit(dbName, query);
      return {id};
    },
  },
  // Map enum GraphQL values to data model required values

  // field-level query
  ComputingDeviceAsset: {
    // installed_hardware: async ( parent, args, context, ) => {
    //   let iriArray = parent.installed_hw_iri;
    // },
    installed_software: async ( parent, args, context, ) => {
      let iriArray = parent.installed_sw_iri;
      var reducer = getSoftwareReducer('SOFTWARE-IRI');
      if (Array.isArray(iriArray) && iriArray.length > 0) {
        const results = [];
        for (let item of iriArray) {
          // check if this is an IPv4 object
          if (!item.includes('Software')) {
            continue;
          }

          // query for the IP address based on its IRI
          var sparqlQuery = getSoftwareQuery('SOFTWARE-IRI', item);
          const response = await context.dataSources.Stardog.queryById( context.dbName, sparqlQuery, singularizeSchema )
          if (response === undefined ) return [];
          if (response && response.length > 0) {
            results.push(reducer( response[0] ))
          }
        }

      return results;
      } else {
        return [];
      }
    },
    installed_operating_system: async ( parent, args, context, ) => {
      var iri = parent.installed_os_iri
      if (Array.isArray( iri ) && iri.length > 0) {
        if (iri.length > 1) {
          console.log(`[WARNING] ${parent.parent_iri} has ${parent.parent_iri.length} values: ${parent.installed_os_iri}`);
          iri = parent.installed_os_iri[0]
        }
      } else {
        iri = parent.installed_os_iri;
      }
      var sparqlQuery = getSoftwareQuery('OS-IRI', iri);
      var reducer = getSoftwareReducer('OS-IRI');
      const response = await context.dataSources.Stardog.queryById( context.dbName, sparqlQuery, singularizeSchema )
      if (response === undefined ) return null;
      if (response && response.length > 0) {
        // console.log( response[0] );
        let results = reducer(response[0])
        return results
      }
    },
    ipv4_address: async ( parent, args, context, ) => {
      let iriArray = parent.ip_addr_iri;
      var reducer = getReducer('IPV4-ADDR');
      if (Array.isArray(iriArray) && iriArray.length > 0) {
        const results = [];
        for (let ipAddr of iriArray) {
          // check if this is an IPv4 object
          if (!ipAddr.includes('IpV4Address')) {
            continue;
          }

          // query for the IP address based on its IRI
          var sparqlQuery = getSelectSparqlQuery('IPV4-ADDR', ipAddr);
          let response;
          try {
            response = await context.dataSources.Stardog.queryById( context.dbName, sparqlQuery, singularizeSchema )
          }catch(e) {
            console.log(e)
            throw e
          }
          if (response === undefined ) return [];
          if (response && response.length > 0) {
            results.push(reducer( response[0] ))
          }
        }

      return results;
      } else {
        return [];
      }
    },
    ipv6_address: async ( parent, args, context, ) => {
      let iriArray = parent.ip_addr_iri;
      var reducer = getReducer('IPV6-ADDR');
      if (Array.isArray(iriArray) && iriArray.length > 0) {
        const results = [];
        for (let ipAddr of iriArray) {
          // check if this is an IPv6 object
          if (!ipAddr.includes('IpV6Address')) {
            continue;
          }

          // query for the IP address based on its IRI
          var sparqlQuery = getSelectSparqlQuery('IPV6-ADDR', ipAddr);
          const response = await context.dataSources.Stardog.queryById( context.dbName, sparqlQuery, singularizeSchema )
          if (response === undefined ) return [];
          if (response.length > 0 ) {
            results.push(reducer( response[0] ))
          }
        }

        return results;
        } else {
          return [];
        }
    },
    mac_address: async ( parent, args, context,) => {
      let iriArray = parent.mac_addr_iri;
      var reducer = getReducer('MAC-ADDR');
      if (Array.isArray(iriArray) && iriArray.length > 0) {
        const results = [];
        const value_array = [];
        for (let addr of iriArray) {
          // check if this is an MAC address object
          if (!addr.includes('MACAddress')) {
            continue;
          }

          // query for the MAC address based on its IRI
          var sparqlQuery = getSelectSparqlQuery('MAC-ADDR', addr);
          let response;
          try {
            response = await context.dataSources.Stardog.queryById( context.dbName, sparqlQuery, singularizeSchema )
          }catch(e) {
            console.log(e)
            throw e
          }
          if (response === undefined ) return [];
          if (response.length > 0) {
            results.push(reducer( response[0] ) )      // TODO: revent back when data is returned as objects, not strings
            // Support for returning MAC address as a string, not a node
            value_array.push( reducer(response[0]).mac_address_value )
          }
        }

        // console.log(`value array: ${value_array}`)
        return value_array
        // return results;      TODO:  revert back when data is returned as objects, not strings
        } else {
          return [];
        }
    },
    ports: async ( parent, args, context, ) => {
      let iriArray = parent.ports_iri;
      var reducer = getReducer('PORT-INFO');
      if (Array.isArray(iriArray) && iriArray.length > 0) {
        const results = [];
        for (let ipAddr of iriArray) {
          // check if this is an IPv4 object
          if (!ipAddr.includes('Port')) {
            continue;
          }

          // query for the IP address based on its IRI
          var sparqlQuery = getSelectSparqlQuery('PORT-INFO', ipAddr);
          const response = await context.dataSources.Stardog.queryById( context.dbName, sparqlQuery, singularizeSchema )
          if (response && response.length > 0) {
            results.push(reducer( response[0] ))
          }
        }

      return results;
      } else {
        return [];
      }
    },
  },
  ComputingDeviceKind: {
    __resolveType: ( item ) => {
      return objectTypeMapping[item.entity_type];
    }
  }
};


export default computingDeviceResolvers;
