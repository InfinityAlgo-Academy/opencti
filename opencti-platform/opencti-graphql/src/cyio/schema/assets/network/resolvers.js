import { assetSingularizeSchema as singularizeSchema } from '../asset-mappings.js';
import {
  getSelectSparqlQuery,
  getReducer,
  insertQuery,
  deleteNetworkAssetQuery
} from './sparql-query.js';
import {compareValues, generateId, DARKLIGHT_NS, updateQuery} from '../../utils.js';
import {
  deleteIpAddressRange,
  deleteIpQuery,
  insertIPAddressRangeQuery,
  insertIPAddressRangeRelationship,
  insertIPQuery,
  selectIPAddressRange
} from "../assetQueries";
import {UserInputError} from "apollo-server-express";
import {addToInventoryQuery} from "../assetUtil";
import {predicateMap} from "./sparql-query";

const networkResolvers = {
  Query: {
    networkAssetList: async ( _, args, context, info ) => {
      var sparqlQuery = getSelectSparqlQuery('NETWORK');
      var reducer = getReducer('NETWORK')
      const response = await context.dataSources.Stardog.queryAll( 
        context.dbName, 
        sparqlQuery,
        singularizeSchema,
        // args.first,       // limit
        // args.offset,      // offset
        args.filter );    // filter
      if (Array.isArray(response) && response.length > 0) {
        // build array of edges
        const edges = [];
        let limit = (args.first === undefined ? response.length : args.first) ;
        let offset = (args.offset === undefined ? 0 : args.offset) ;
        let assetList ;
        if (args.orderedBy !== undefined ) {
          assetList = response.sort(compareValues(args.orderedBy, args.orderMode ));
        } else {
          assetList = response;
        }
        for (let asset of assetList) {
          // skip down past the offset
          if ( offset ) {
            offset--
            continue
          }

          if ( limit ) {
            let edge = {
              cursor: asset.iri,
              node: reducer( asset ),
            }
            edges.push( edge )
            limit--;
          }
        }
        return {
          pageInfo: {
            startCursor: assetList[0].iri,
            endCursor: assetList[assetList.length -1 ].iri,
            hasNextPage: (args.first > assetList.length ? true : false),
            hasPreviousPage: (args.offset > 0 ? true : false),
            globalCount: assetList.length,
          },
          edges: edges,
        }
      } else {
        return ;
      }
    },
    networkAsset: async (_, args, context, info ) => {
      var sparqlQuery = getSelectSparqlQuery("NETWORK",context.selectMap.getNode("networkAsset"), args.id);
      var reducer = getReducer('NETWORK')
      const response = await context.dataSources.Stardog.queryById( context.dbName, sparqlQuery, singularizeSchema )
      if (response === undefined ) return null;
      const first = response[0];
      if (first === undefined) return null;
      const res = reducer( first );
      return res;
    }
  },
  Mutation: {
    createNetworkAsset: async ( _, {input}, context) => {
      const dbName = context.dbName;
      let ipv4RelIri = null, ipv6RelIri = null;
      if(input.network_ipv4_address_range !== undefined) {
        const ipv4Range = input.network_ipv4_address_range;
        delete input.network_ipv4_address_range;
        const {ipIris: startIris, query: startQuery } = insertIPQuery([ipv4Range.starting_ip_address], 4);
        const {ipIris: endIris, query: endQuery } = insertIPQuery([ipv4Range.ending_ip_address], 4);
        const startIri = startIris[0], endIri = endIris[0];
        await context.dataSources.Stardog.create(dbName, startQuery);
        await context.dataSources.Stardog.create(dbName, endQuery);
        const {iri, query} = insertIPAddressRangeQuery(startIri, endIri);
        ipv4RelIri = iri;
        await context.dataSources.Stardog.create(dbName, query);
      }
      if(input.network_ipv6_address_range !== undefined){
        const ipv6Range = input.network_ipv6_address_range;
        delete input.network_ipv6_address_range;
        const {ipIris: startIris, query: startQuery } = insertIPQuery([ipv6Range.starting_ip_address], 6);
        const {ipIris: endIris, query: endQuery } = insertIPQuery([ipv6Range.ending_ip_address], 6);
        const startIri = startIris[0], endIri = endIris[0];
        await context.dataSources.Stardog.create(dbName, startQuery);
        await context.dataSources.Stardog.create(dbName, endQuery);
        const {iri, query} = insertIPAddressRangeQuery(startIri, endIri);
        ipv6RelIri = iri;
        await context.dataSources.Stardog.create(dbName, query);
      }

      const {iri, id, query} = insertQuery(input);
      await context.dataSources.Stardog.create(dbName, query);

      if (ipv4RelIri !== null) {
        const relQuery = insertIPAddressRangeRelationship(iri, ipv4RelIri);
        await context.dataSources.Stardog.create(dbName, relQuery);
      }
      if (ipv6RelIri !== null){
        const relQuery = insertIPAddressRangeRelationship(iri, ipv6RelIri);
        await context.dataSources.Stardog.create(dbName, relQuery);
      }

      const connectQuery = addToInventoryQuery(iri);
      await context.dataSources.Stardog.create(dbName, connectQuery);
      return {id}
    },
    deleteNetworkAsset: async ( _, args, context, info ) => {
      const dbName = context.dbName;
      const sparqlQuery = getSelectSparqlQuery("NETWORK", ["id", "network_address_range"], args.id);
      const response = await context.dataSources.Stardog.queryById(dbName, sparqlQuery, singularizeSchema);
      if(response.length === 0) throw new UserInputError(`Entity does not exists with ID ${args.id}`);
      const reducer = getReducer("NETWORK");
      const asset = reducer(response[0]);
      if(asset.netaddr_range_iri){
        const ipRangeQuery = selectIPAddressRange(`<${asset.netaddr_range_iri}>`);
        const ipRange = await context.dataSources.Stardog.queryAll(dbName, ipRangeQuery);
        if(ipRange.length === 1){
          const start = ipRange[0].starting_ip_address;
          const end = ipRange[0].ending_ip_address;
          let ipQuery = deleteIpQuery(`<${start}>`);
          await context.dataSources.Stardog.delete(dbName, ipQuery);
          ipQuery = deleteIpQuery(`<${end}>`);
          await context.dataSources.Stardog.delete(dbName, ipQuery);
        }
        const deleteIpRange = deleteIpAddressRange(`<http://scap.nist.gov/ns/asset-identification#IpAddressRange-${args.id}>`);
        await context.dataSources.Stardog.delete(dbName, deleteIpRange);
      }
      const deleteQuery = deleteNetworkAssetQuery(args.id);
      await context.dataSources.Stardog.delete(dbName, deleteQuery);
      return id
    },
    editNetworkAsset: async ( _, {id, input}, context ) => {
      const dbName = context.dbName;
      const query = updateQuery(
          `http://scap.nist.gov/ns/asset-identification#Network-${id}`,
          "http://scap.nist.gov/ns/asset-identification#Network",
          input,
          predicateMap
      )
      await context.dataSources.Stardog.edit(dbName, query);
      return {id}
    },
  },
  // Map enum GraphQL values to data model required values
  NetworkAsset: {
    network_address_range: async (parent, args, context, info) => {
      let item = parent.netaddr_range_iri;
      var sparqlQuery = selectIPAddressRange(`<${item}>`)
      var reducer = getReducer('NETADDR-RANGE');
      const response = await context.dataSources.Stardog.queryById( context.dbName, sparqlQuery, singularizeSchema )
      if (response && response.length > 0) {
        // console.log( response[0] );
        // let results = ipAddrRangeReducer( response[0] )    TODO: revert when data is passed as objects, instead of string
        let results = reducer( response[0] )
        let x = generateId( {"value": results.start_addr_iri}, DARKLIGHT_NS)
        return {
          id: results.id,
          starting_ip_address: {
            id: generateId( {"value": results.start_addr_iri}, DARKLIGHT_NS),
            entity_type: (results.start_addr_iri.includes(':') ? 'ipv6-addr' : 'ipv4-addr'),
            ip_address_value: results.start_addr_iri
          },
          ending_ip_address: {
            id: generateId( {"value": results.ending_addr_iri}, DARKLIGHT_NS),
            entity_type: (results.ending_addr_iri.includes(':') ? 'ipv6-addr' : 'ipv4-addr'),
            ip_address_value: results.ending_addr_iri
          }
        }
        // return results
      }
    }
  }
};

export default networkResolvers;
