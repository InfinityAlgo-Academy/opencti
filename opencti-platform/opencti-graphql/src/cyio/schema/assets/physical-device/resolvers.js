const physicalDeviceResolvers = {
  Query: {},
  Mutation: {},
  // Map enum GraphQL values to data model required values
};

export default physicalDeviceResolvers;
