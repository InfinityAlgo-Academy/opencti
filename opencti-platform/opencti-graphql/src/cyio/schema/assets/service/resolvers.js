const serviceResolvers = {
    Query: {
  
    },
    Mutation: {
  
    },
    // Map enum GraphQL values to data model required values
  
  
  };
  
  export default serviceResolvers;