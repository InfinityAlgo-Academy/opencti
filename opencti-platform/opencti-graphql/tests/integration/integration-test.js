import { createTestClient } from 'apollo-server-testing';
import { ApolloServer } from 'apollo-server-express';
import schema from '../../src/schema/schema';
import { initializeSchema } from '../../src/initialization';
import applyMigration from '../../src/database/migration';

// Initialize schema before tests
beforeAll(async () => {
  await initializeSchema();
  return applyMigration();
}, 120000);

// Setup the configuration
export const USER_ID = 'V1234';
const server = new ApolloServer({
  schema,
  context: () => ({
    user: {
      id: USER_ID,
      name: 'admin',
      email: 'admin@opencti.io',
      grant: ['ROLE_ADMIN', 'ROLE_ROOT']
    }
  })
});
export const { query } = createTestClient(server);
jest.setTimeout(40000);

// Starting tests
require('./database/grakn');
require('./resolvers/user');
require('./resolvers/threatActor');
