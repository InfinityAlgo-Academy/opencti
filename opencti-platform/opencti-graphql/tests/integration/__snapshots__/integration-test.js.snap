// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User resolver standard behavior should user created 1`] = `
Object {
  "data": Object {
    "userAdd": Object {
      "user_email": "admin@opencti.io",
      "firstname": "Administrator",
      "lastname": "OpenCTI",
      "name": "admin",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
